sql/catalog/externalcatalog: TestExtractIngestExternalCatalog failed

sql/catalog/externalcatalog.TestExtractIngestExternalCatalog [failed](https://tanzanite.cluster.engflow.com/invocations/default/7ec3784f-893a-4ddb-9cc4-1b7cf3da080e?testReportRun=2&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NhdGFsb2cvZXh0ZXJuYWxjYXRhbG9nOmV4dGVybmFsY2F0YWxvZ190ZXN0) on master @ [1aba4038242990fdb37a5317512fc5e82d6334fe](https://github.com/cockroachdb/cockroach/commits/1aba4038242990fdb37a5317512fc5e82d6334fe):


```
=== RUN   TestExtractIngestExternalCatalog
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestExtractIngestExternalCatalog3943006934
    test_log_scope.go:76: use -show-logs to present logs inline
    external_catalog_test.go:140: -- test log scope end --
test logs left over in: outputs.zip/logTestExtractIngestExternalCatalog3943006934
--- FAIL: TestExtractIngestExternalCatalog (17.65s)
=== RUN   TestExtractIngestExternalCatalog/basic
    sql_runner.go:298: error executing 'SELECT name FROM crdb_internal.tables WHERE state = 'DROP'': pq: referenced descriptor ID 109: looking up ID 109: descriptor not found
    --- FAIL: TestExtractIngestExternalCatalog/basic (3.06s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExtractIngestExternalCatalog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-46176

sql/catalog/externalcatalog.TestExtractIngestExternalCatalog [failed](https://tanzanite.cluster.engflow.com/invocations/default/c6fbd54c-f28d-448c-877b-21f2d1a0af85?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NhdGFsb2cvZXh0ZXJuYWxjYXRhbG9nOmV4dGVybmFsY2F0YWxvZ190ZXN0) on master @ [dc789b58edb67ae00f16117e2daee179fedc4a19](https://github.com/cockroachdb/cockroach/commits/dc789b58edb67ae00f16117e2daee179fedc4a19):


```
=== RUN   TestExtractIngestExternalCatalog
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestExtractIngestExternalCatalog747474018
    test_log_scope.go:76: use -show-logs to present logs inline
    external_catalog_test.go:140: -- test log scope end --
test logs left over in: outputs.zip/logTestExtractIngestExternalCatalog747474018
--- FAIL: TestExtractIngestExternalCatalog (17.06s)
=== RUN   TestExtractIngestExternalCatalog/basic
    sql_runner.go:298: error executing 'SELECT name FROM crdb_internal.tables WHERE state = 'DROP'': pq: referenced descriptor ID 109: looking up ID 109: descriptor not found
    --- FAIL: TestExtractIngestExternalCatalog/basic (2.74s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExtractIngestExternalCatalog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/catalog/externalcatalog/external_catalog_test.go
+++ b/pkg/sql/catalog/externalcatalog/external_catalog_test.go
@@ -23,6 +23,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/redact"
 	"github.com/stretchr/testify/require"
 )
@@ -48,7 +49,12 @@ func TestExtractIngestExternalCatalog(t *testing.T) {
 	sysDB := sqlutils.MakeSQLRunner(srv.SystemLayer().SQLConn(t))
 	sysDB.Exec(t, "SET CLUSTER SETTING sql.virtual_cluster.feature_access.zone_configs.enabled = true;")
 	sysDB.Exec(t, "SET CLUSTER SETTING sql.virtual_cluster.feature_access.zone_configs_unrestricted.enabled = true;")
-	sysDB.Exec(t, "SET CLUSTER SETTING sql.gc_job.wait_for_gc.interval = '250ms'")
+
+	rng, _ := randutil.NewTestRand()
+	fastGC := rng.Float64() < 0.5
+	if fastGC {
+		sysDB.Exec(t, "SET CLUSTER SETTING sql.gc_job.wait_for_gc.interval = '250ms'")
+	}
 
 	sqlDB := sqlutils.MakeSQLRunner(conn)
 	sqlDB.Exec(t, "CREATE DATABASE db1")
@@ -113,12 +119,17 @@ func TestExtractIngestExternalCatalog(t *testing.T) {
 		var res int
 		sqlDB.QueryRow(t, "SELECT count(*) FROM [SHOW TABLES]").Scan(&res)
 		require.Zero(t, res)
-		sqlDB.CheckQueryResults(t, "SELECT name FROM crdb_internal.tables WHERE state = 'DROP'", [][]string{{"tab1"}, {"tab2"}})
-		sqlDB.CheckQueryResultsRetry(
-			t,
-			"SELECT status FROM [SHOW JOBS] WHERE description = 'test gc'",
-			[][]string{{"succeeded"}},
-		)
+		if fastGC {
+			// With fast gc, the dropped tables may disappear before
+			// crdb_internal.tables is read.
+			sqlDB.CheckQueryResultsRetry(
+				t,
+				"SELECT status FROM [SHOW JOBS] WHERE description = 'test gc'",
+				[][]string{{"succeeded"}},
+			)
+		} else {
+			sqlDB.CheckQueryResults(t, "SELECT name FROM crdb_internal.tables WHERE state = 'DROP'", [][]string{{"tab1"}, {"tab2"}})
+		}
 	})
 
 	t.Run("fk", func(t *testing.T) {

